### PR TITLE
Fix logits processor when prompt_logprobs is not None

### PR DESCRIPTION
**Pull request by huyiwen**
Hello,

I am currently developing an LLM toolkit based on vLLM. However, I encountered a bug related to logits processor #2141 developed by @noamgat within vllm detailed at #2800 . The primary issue with this bug is that when `prompt_logprobs` in `SamplingParams` is set to a value other than None, the logits for the first iteration include the prompt. This evidently occurs while calculating the logprobs for the tokens corresponding to the prompt. However, this leads to an AssertionError at [this](https://github.com/vllm-project/vllm/blob/f7c1234990793008f3d44790fd274040f26c4ee4/vllm/model_executor/layers/sampler.py#L162) assert statement.

I have fixed this bug when `prompt_logprobs` is not None and conducted tests with different values for `prompt_logprobs`. With these changes, the error does not occur in either scenario.

In detail, when `len(sampling_metadata.prompt_lens) > 0`, it means this is the first iteration; whereas `sum(sampling_metadata.prompt_lens) == logits.shape[0]` implies that the logits include the prompt.

Thank you for your review. I look forward to your fix for this issue, which will assist us in completing our project development.

Wishing you a happy day.

### CODE REVIEW

**Comment by huyiwen** [comment]
I found that I cited an incorrect pull request. Logits processor is introduced here: #1469 

@simon-mo may be able to review this pr.

**Comment by tjohnson31415** [comment]
Hello @huyiwen. I found your PR after also encountering the AssertionError and working on an independent fix in https://github.com/vllm-project/vllm/pull/3867. I was trying to actually process the prompt logits with logits_processors, but decided that it would be easier to just skip them like in your PR 😅 So I'd like to help get this merged in if I can.

I think there is another case to handle in your PR:
if there are multiple requests batched together where only a subset have `prompt_logprobs`, then we need to check each request independently to decide to skip instead of skipping for all prompts in the batch (only the request with `prompt_logprobs` will have rows in `logits` for each prompt token).

<details>
<summary>Script to reproduce</summary>

```
import torch
from vllm import LLM, SamplingParams

model_path = 'facebook/opt-125m'

llm = LLM(model=model_path)

prompts = [
    "Hello, My name is",
]

def pick_ith(token_ids, logits):
    # skip lower ids that have special tokens
    logits[len(token_ids)+10] = torch.finfo(torch.float16).max
    return logits
    
sampling_params = SamplingParams(
    temperature=0.5,
    logits_processors=[pick_ith],
    prompt_logprobs=3,
)

prompt = "Hello, My name is"

llm._add_request(
    prompt = prompt,
    sampling_params = sampling_params,
    prompt_token_ids = None
)
llm._add_request(
    prompt = prompt,
    sampling_params = SamplingParams(), # default sampling_params without prompt_logprobs
    prompt_token_ids = None
)

outputs = llm._run_engine(False)

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
 ```
 
</details>

File: vllm/sampling_params.py [hunk 1]
```diff
@@ -261,4 +261,6 @@ def __repr__(self) -> str:
             f"prompt_logprobs={self.prompt_logprobs}, "
             f"skip_special_tokens={self.skip_special_tokens}, "
             "spaces_between_special_tokens="
-            f"{self.spaces_between_special_tokens})")
+            f"{self.spaces_between_special_tokens}, "
+            "logits_processors="
+            f"{self.logits_processors})")
```
Comment (njhill) [inline vllm/sampling_params.py:266]: Could you remove this change from the PR? It would be good to handle in a separate PR since I'm not sure we necessarily want to log the string form of all of the logits processors (which could be arbitrary).

**Review by njhill — COMMENTED** [review]
@huyiwen thanks for this fix! We're be interested in getting it merged asap. Would you be able to address @tjohnson31415's comment? You can see an example of a similar fix that he's made for the min_tokens processing [here](https://github.com/vllm-project/vllm/pull/3876).

**Comment by huyiwen** [comment]
It seems _apply_logits_processors has been moved to a different file. I've opened a new PR to accommodate it: #3899.
